source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc

```diff
@@ -0,0 +1,60 @@
+#include "extensions/filters/http/dynamic_forward_proxy/proxy_filter.h"
+
+#include "extensions/common/dynamic_forward_proxy/dns_cache.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace DynamicForwardProxy {
+
+ProxyFilterConfig::ProxyFilterConfig(
+    const envoy::config::filter::http::dynamic_forward_proxy::v2alpha::FilterConfig& proto_config,
+    Extensions::Common::DynamicForwardProxy::DnsCacheManagerFactory& cache_manager_factory,
+    Upstream::ClusterManager& cluster_manager)
+    : dns_cache_manager_(cache_manager_factory.get()),
+      dns_cache_(dns_cache_manager_->getCache(proto_config.dns_cache_config())),
+      cluster_manager_(cluster_manager) {}
+
+void ProxyFilter::onDestroy() {
+  // Make sure we destroy any active cache load handle in case we are getting reset and deferred
+  // deleted.
+  cache_load_handle_.reset();
+}
+
+Http::FilterHeadersStatus ProxyFilter::decodeHeaders(Http::HeaderMap& headers, bool) {
+  Router::RouteConstSharedPtr route = decoder_callbacks_->route();
+  if (!route || !route->routeEntry()) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  Upstream::ThreadLocalCluster* cluster =
+      config_->clusterManager().get(route->routeEntry()->clusterName());
+  if (!cluster) {
```

Comment:
what are the semantics of activating the filter on a cluster of different type, or different DNS cache?
is there a way to check if the cluster's dns cache has the same name as the filter? 
it sounds like the filter should Continue (if we assume multiple instances in the chain) or Error\Assert (if not) if it is not the case?

Reply:
I thought about this, and there is no perfect solution to notify the user if they messed up the configuration, especially in light of CDS/LDS, etc. Let me add a TODO/comment here to see if we can think about a better way to handle this.